### ISSUE TITLE
`Condition::default()` creates an invalid condition that can't be used

### ISSUE THREAD

**Issue by daniel-abramov** [qdrant/rust-client#205]
The conditions created by `Condition::default()` are [rejected by Qdrant](https://github.com/qdrant/qdrant/blob/27260abda78509e1a3e8822c8d8819c4fe189f5b/lib/api/src/grpc/conversions.rs#L1082) backend with a "Malformed Condition type" error.

**Comment by timvisee** [comment]
Thanks!

I think we should remove the `Default` implementation on `Condition` to prevent misuse of this with provided functions.

I'm sure you are aware, but in the mean time you can [one](https://docs.rs/qdrant-client/latest/qdrant_client/qdrant/struct.Condition.html#method.matches) [of](https://docs.rs/qdrant-client/latest/qdrant_client/qdrant/struct.Condition.html#method.matches_text) [the](https://docs.rs/qdrant-client/latest/qdrant_client/qdrant/struct.Condition.html#method.range) [other](https://docs.rs/qdrant-client/latest/qdrant_client/qdrant/struct.Condition.html#method.datetime_range) [functions](https://docs.rs/qdrant-client/latest/qdrant_client/qdrant/struct.Condition.html#method.nested) to create the desired condition.

**Comment by timvisee** [comment]
After quick inspection - I'm afraid we cannot remove the `Default` implementation at this time because gRPC framework we use puts it there for us.

It is implied by `::post::Message`: https://github.com/qdrant/rust-client/blob/1707b92b0bbfb2f6a16bc7d34d579ad9fb8eff04/src/qdrant.rs#L5636

**Comment by daniel-abramov** [comment]
> I'm sure you are aware, but in the mean time you can [one](https://docs.rs/qdrant-client/latest/qdrant_client/qdrant/struct.Condition.html#method.matches) [of](https://docs.rs/qdrant-client/latest/qdrant_client/qdrant/struct.Condition.html#method.matches_text) [the](https://docs.rs/qdrant-client/latest/qdrant_client/qdrant/struct.Condition.html#method.range) [other](https://docs.rs/qdrant-client/latest/qdrant_client/qdrant/struct.Condition.html#method.datetime_range) [functions](https://docs.rs/qdrant-client/latest/qdrant_client/qdrant/struct.Condition.html#method.nested) to create the desired condition.

Sure. In my case, the way it occurred is that I had a function that received a structure with optional fields, which I then converted into conditions that were passed to the `Filter`, e.g.:

```rust
let city_filter = data.location.map(|loc| Condition::matches("city", loc)).unwrap_or_default();
let radius_filter = data.radius.map(|r| Condition::geo_radius("radius", r.into())).unwrap_or_default();
let filter = Filter::all(vec![city_filter, radius_filter]);
```

So it was a compact, elegant implementation that I assumed would work (typically, if there is a `Default` implementation, it produces a structure that is valid and can be used), but later I discovered that the function fails in some cases and traced it back to that place in Qdrant that rejects `None` conditions.

> After quick inspection - I'm afraid we cannot remove the Default implementation at this time because gRPC framework we use puts it there for us.

There are still several ways to mitigate it. For instance, instead of exposing the low-level "over-the-wire" representation of the `Condition`, one could expose a convenient high-level type to the user and only convert it to the low-level representations inside a library. This ensures that the client only interacts with a "safe" high-level API that prevents errors.

Alternatively, there is a "lazy solution", such as making the functions that accept `Condition`, such as `Filter::all` or other, to return an error if they receive an invalid condition. It would still be more efficient than letting the user to send an **invalid request** to the server, only to discover that it's not valid and gets rejected, i.e. we could spare that additional round-trip and return error immediately when we know upfront that the request is invalid and won't be processed anyway.

However, a better design would simply prevent creating "invalid" structures to avoid bugs and issues 🙂 And to be honest, the `Condition` type is **almost** perfect already - it provides lots of useful high-level functions, that's why I think that it just needs to either not have `Default`, or make `Default` return something that is valid. If its internal fields could be misused, they should also be made private. I.e. I think there is a simple way to fix the issue by just wrapping the internal 'over-the-wire' `Condition` into a newtype that prevents such things from happening.

**Comment by timvisee** [comment]
Yes, I totally agree. I understand that it's easy to construct this 'broken' variant the way you describe. Thanks for all the suggestions.

I implemented a quick fix to filter out empty conditions, resolving this issue: https://github.com/qdrant/qdrant/pull/5690

**Comment by timvisee** [comment]
We've just released [Qdrant 1.12.6](https://github.com/qdrant/qdrant/releases/tag/v1.12.6) which includes the patch for this problem. Thank again for bringing this to our attention!
